lib/stripe/credit_card_cloner.rb

```diff
@@ -1,38 +1,115 @@
 #
 # This is required when using the Stripe Payment Intents API:
 #   - the customer and payment methods are stored in the platform account
 #       so that they can be re-used across multiple sellers
-#   - when a card needs to be charged, we need to create it in the seller's stripe account
+#   - when a card needs to be charged, we need to clone (or find the clone)
+#       in the seller's stripe account
 #
-# We are doing this process every time the card is charged:
-#   - this means that, if the customer uses the same card on the same seller multiple times,
-#       the card will be created multiple times on the seller's account
-#   - to avoid this, we would have to store the IDs of every card on each seller's stripe account
-#       in our database (this way we only have to store the platform account ID)
+# To avoid creating a new clone of the card/customer each time the card is charged or
+# authorized (e.g. for SCA), we attach metadata { clone: true } to the card the first time we
+# clone it and look for a card with the same fingerprint (hash of the card number) and
+# that metadata key to avoid cloning it multiple times.
+
 module Stripe
   class CreditCardCloner
+    def find_or_clone(card, connected_account_id)
+      if card.user && cloned_card = find_cloned_card(card, connected_account_id)
+        cloned_card
+      else
+        clone(card, connected_account_id)
+      end
+    end
+
+    def destroy_clones(card)
+      card.user.customers.each do |customer|
+        next unless stripe_account = customer.enterprise.stripe_account&.stripe_user_id
+
+        customer_id, _payment_method_id = find_cloned_card(card, stripe_account)
+        next unless customer_id
+
+        customer = Stripe::Customer.retrieve(customer_id, stripe_account: stripe_account)
+        customer&.delete unless customer.deleted?
+      end
+    end
+
+    private
+
     def clone(credit_card, connected_account_id)
       new_payment_method = clone_payment_method(credit_card, connected_account_id)
 
       # If no customer is given, it will clone the payment method only
-      return nil, new_payment_method.id if credit_card.gateway_customer_profile_id.blank?
+      return [nil, new_payment_method.id] if credit_card.gateway_customer_profile_id.blank?
 
       new_customer = Stripe::Customer.create({ email: credit_card.user.email },
                                              stripe_account: connected_account_id)
       attach_payment_method_to_customer(new_payment_method.id,
                                         new_customer.id,
                                         connected_account_id)
 
+      add_metadata_to_payment_method(new_payment_method.id, connected_account_id)
+
       [new_customer.id, new_payment_method.id]
     end
 
-    private
+    def find_cloned_card(card, connected_account_id)
+      return nil unless fingerprint = fingerprint_for_card(card)
+
+      find_customers(card.user.email, connected_account_id).each do |customer|
```

Comment:
I'm a bit concerned about the performance of this. The list of customers can potentially be large and it'll grow over time. Now, let's add network requests into the mix and this will slow to a crawl. Not to mention the unreliability of the network, so the more we perform, the more changes of experiencing a failure.

Is there a chance to pass the stripe customer id into this class so we can fetch only that one?

Reply:
It's because of what we've been doing where we clone the card (and customer) from the platform account to the connected account each time we need to take payment or authorize. Each of those cloned customers has the same email address so we need to loop through them all to find a cloned card.

Reply:
and there's no way we could migrate that data as we would do with a DB data migration? probably we can think about it after releasing this.

Reply:
We could think about a migration, but we'd have to be careful. I thought about having the method delete any cards that looked like clones (i.e. had the same fingerprint and the same email address for the customer), but on the off chance that the shop had manually added a customer/card with matching info, it seemed like we would delete that too, so I didn't want to risk messing with a shop's data. It still might be possible with some additional queries, but not easy.